sphinx/ext/doctest.py

```diff
@@ -394,13 +396,53 @@ def condition(node):
                 return isinstance(node, (nodes.literal_block, nodes.comment)) \
                     and 'testnodetype' in node
         for node in doctree.traverse(condition):
+            try:
+                # Try to get the file which actually contains the doctest,
+                # not the filename of the document it's included in.
+                filename = path.relpath(node.source, self.env.srcdir)\
+                    .replace(path.sep, '/')\
+                    .rsplit(':docstring of ', maxsplit=1)[0]
+            except Exception:
+                filename = self.env.doc2path(docname, base=None)
             source = 'test' in node and node['test'] or node.astext()
+            # HACK HACK HACK - `node.line` has an off by one error somewhere
+            line_number = node.line - 1 if node.line is not None else None
+            if ':docstring of ' in path.basename(node.source or ''):
```

Comment:
I feel this block is a bit large. How about splitting into a function like following?
```
def get_source_lineno(node, source, cache={}):
    ...
```

I think main block should be simple and readable.